csharp/src/Google.Protobuf/Reflection/FileDescriptor.cs

```diff
@@ -445,9 +446,9 @@ private void CrossLink()
             }
         }
 
-        private static void AddAllExtensions(FileDescriptor[] dependencies, GeneratedClrTypeInfo generatedInfo, ExtensionRegistry registry)
+        private static IEnumerable<Extension> GetAllExtensions(FileDescriptor[] dependencies, GeneratedClrTypeInfo generatedInfo)
         {
-            registry.AddRange(dependencies.SelectMany(GetAllDependedExtensions).Concat(GetAllGeneratedExtensions(generatedInfo)).ToArray());
+            return dependencies.SelectMany(GetAllDependedExtensions).Distinct().Concat(GetAllGeneratedExtensions(generatedInfo));
```

Comment:
Should we get an enumerable over the files and run distinct on those so we don't duplicate work by loading the file's extensions multiple times?